data/src/main/java/com/microsoft/azure/kusto/data/exceptions/DataServiceException.java

```diff
@@ -3,26 +3,30 @@
 
 package com.microsoft.azure.kusto.data.exceptions;
 
-public class DataServiceException extends Exception {
+public class DataServiceException extends KustoClientException {
     private String ingestionSource;
-
+    private TriState isPermanent;
     public String getIngestionSource() { return ingestionSource; }
 
     public DataServiceException(String message) {
-        super(message);
+        this(null, message, TriState.DONT_KNOW);
     }
 
-    public DataServiceException(String message, Exception exception) {
-        super(message, exception);
+    public DataServiceException(String message, Exception exception, TriState isPermanent) {
+        this(null, message, exception, isPermanent);
     }
 
-    public DataServiceException(String ingestionSource, String message, Exception exception) {
+    public DataServiceException(String ingestionSource, String message, TriState isPermanent) {
+        this(ingestionSource, message, null, isPermanent);
+    }
+
+    public DataServiceException(String ingestionSource, String message, Exception exception, TriState isPermanent) {
         super(message, exception);
         this.ingestionSource = ingestionSource;
+        this.isPermanent = isPermanent;
     }
 
-    public DataServiceException(String ingestionSource, String message) {
-        super(message);
-        this.ingestionSource = ingestionSource;
+    public TriState isPermanent(){
```

Comment:
This can inherit from its parent if it has the concrete implementation I suggested.